Depth-limited printing of types in stacktraces take 2!

JeffBezanson opened the pull request:
This is an alternate implementation of #49231. It's the same idea but I implemented it just by traversing the output string, counting curly braces, and omitting text at the needed nesting depth. It handles escape sequences by skipping over them. @timholy can you poke any holes in this? I think it should work in practice and is much less code. If no fatal problems are found I'll port over the tests.

JeffBezanson committed:
limit printing depth of argument types in stack traces

Co-authored-by: Tim Holy <tim.holy@gmail.com>

timholy commented:
Thanks! I haven't looked in detail, but the principle of doing more things with raw strings is a good one. I'll be happy to give it a thorough review after one issue worthy of discussion: one not-so-obvious goal of my previous implementation was to create an object that could, if desired, be passed to [FoldingTrees](https://github.com/JuliaCollections/FoldingTrees.jl) for interactive exploration of the type. But I am unsure that's a sufficiently important goal to be worth a far more complicated implementation. What do you think?

vtjnash commented on base/show.jl:
```diff
@@ -2470,15 +2470,23 @@ function print_within_stacktrace(io, s...; color=:normal, bold=false)
         return
     end
     tv = Any[]
+    if get(out, :limit, false)::Bool
+        io = IOBuffer()
+        if isa(out, IOContext)
+            io = IOContext(io, out)
+        end
+    else
+        io = out
+    end
```
Type-unstable code like this should generally be avoided. There is no reason to not call the `IOContext` wrapper unconditionally, and it is incorrect anyways not to wrap it when making an IOBuffer that is supposed to act like the original IO (several properties still get copied over, even if it is not previously an IOContext).

vtjnash commented on base/show.jl:
```diff
@@ -2516,9 +2524,58 @@ function show_tuple_as_call(io::IO, name::Symbol, sig::Type;
+        end
+        if c == '{'
+            depth += 1
+        elseif in_escape && c == 'm'
```
Why is it safe to assume ANSI parsing and that `m` is the only valid ANSI code for generic IO?

JeffBezanson replied:
Oh excellent, I didn't realize we had `ANSIIterator` for this!

vtjnash commented on base/show.jl:
```diff
@@ -2516,9 +2524,58 @@ function show_tuple_as_call(io::IO, name::Symbol, sig::Type;
+        elseif in_escape && c == 'm'
+            in_escape = false
+        end
+    end
```
This would fail on a type such as:
```
julia> Val{'{'}
Val{'{'}
```

vtjnash replied:
This might be better after https://github.com/JuliaLang/julia/pull/49586, since making this reliable and general is the point of that PR IIUC (similar to https://github.com/JuliaLang/julia/pull/27430 which also had that, but which I never managed to finish)

timholy reviewed (approved):
Overall I really like this. I think we could just solve the stacktrace printing problem and decide later if we want to be able to introspect on types interactively.

It seems like one or perhaps two changes might be necessary:
- detect when braces are not syntactically meaningful (@vtjnash's `Val{'{'}` and also `Val{Symbol("2^{x-y}")}`), which I think we could handle with `is_quoted` similar to `in_escape`
- possibly more nuanced handling of escape codes. Since this is only used for printing types---and we tell people not to specialize this---I think the world of possible escapes is limited. If we merge #49586 that would provide one way of handling it; if you need something that does slightly more validation, you could steal anything you want from https://github.com/JuliaIO/WidthLimitedIO.jl/blob/main/src/ansi_escapes.jl

I'd approve porting over the tests I included and merging this once other concerns are addressed. I'm not sure my PR included the full suite of tests I had in mind, but what I was struggling with (reproducing the character-for-character coloring when the width of the display was not an issue) may not be applicable with the approach taken here.

JeffBezanson commented:
OK I think this is ready to go. I used the lovely `ANSIIterator` to skip escape sequences more accurately, and handled basic single and double quoting. I added some tests for those cases plus color output. Also added the hint for how to see full information when types are truncated.

vtjnash commented on test/stacktraces.jl:
```diff
@@ -192,3 +192,57 @@ let bt
+    end
+    str = sprint(Base.show_backtrace, st, context = (:limit=>true, :color=>true, :displaysize=>(50,132)))
+    @test contains(str, "[2] \e[0m\e[1m(::$F49231{Vector, Val{…}, Vector{…}, NTuple{…}, Int64, Int64, Int64})\e[22m\e[0m\e[1m(\e[22m\e[90ma\e[39m::\e[0mInt64, \e[90mb\e[39m::\e[0mInt64, \e[90mc\e[39m::\e[0mInt64\e[0m\e[1m)\e[22m\n\e[90m")
+end
```
If you need a test:
```
julia> Val{:var"\\\""}
Val{Symbol("\\\"")}
```

timholy commented on base/show.jl:
```diff
@@ -2516,9 +2517,117 @@ function show_tuple_as_call(io::IO, name::Symbol, sig::Type;
+            depth += 1
+            if depth > length(width_at)
+                resize!(width_at, depth)
+                width_at[depth] = 0
```
If it grows one-at-a-time (seems like it should), isn't this equivalent to `push!`? If not, doesn't this risk leaving the elements between `width_at[oldlen+1:end-1]` uninitialized?

vtjnash commented on base/show.jl:
```diff
@@ -2516,9 +2517,117 @@ function show_tuple_as_call(io::IO, name::Symbol, sig::Type;
+    str = String(take!(unwrapcontext(io)[1]))
+    if get(out, :limit, false)::Bool
+        sz = get(out, :displaysize, (typemax(Int), typemax(Int)))::Tuple{Int, Int}
+        str_lim = type_depth_limit(str, sz[2])
```
Can we make this min(width,120), so that very skinny terminals still get a reasonable amount of output (particularly for copy paste)?

timholy replied:
If the concern is skinny terminals this should be `max(width, 120)` (not `min`)...but it seems like Jeff caught that.

charleskawczynski commented:
How can users use this non-interactively?
